addons/mail/models/fetchmail.py

```diff
@@ -190,102 +238,101 @@ def button_confirm_login(self):
+
+        Commit after each message.
+        """
+        result_exception = None
+        servers = self.with_context(fetchmail_cron_running=True)
+        total_remaining = len(servers)  # number of remaining messages + number of unchecked servers
+        self.env['ir.cron']._commit_progress(remaining=total_remaining)
+
+        for server in servers:
+            total_remaining -= 1  # the server is checked
+            if not server.try_lock_for_update(allow_referencing=True).filtered_domain(MAIL_SERVER_DOMAIN):
+                _logger.info('Skip checking for new mails on mail server id %d (unavailable)', server.id)
+                continue
+            server_type_and_name = server.server_type, server.name  # avoid reading this after each commit
+            _logger.info('Start checking for new emails on %s server %s', *server_type_and_name)
             count, failed = 0, 0
-            imap_server = None
-            pop_server = None
-            connection_type = server._get_connection_type()
-            if connection_type == 'imap':
-                try:
-                    imap_server = server.connect()
-                    imap_server.select()
-                    result, data = imap_server.search(None, '(UNSEEN)')
-                    for num in data[0].split():
-                        res_id = None
-                        result, data = imap_server.fetch(num, '(RFC822)')
-                        imap_server.store(num, '-FLAGS', '\\Seen')
-                        try:
-                            res_id = MailThread.with_context(**additionnal_context).message_process(server.object_id.model, data[0][1], save_original=server.original, strip_attachments=(not server.attach))
-                        except Exception:
-                            _logger.info('Failed to process mail from %s server %s.', server.server_type, server.name, exc_info=True)
-                            failed += 1
-                        imap_server.store(num, '+FLAGS', '\\Seen')
-                        self._cr.commit()
-                        count += 1
-                    _logger.info("Fetched %d email(s) on %s server %s; %d succeeded, %d failed.", count, server.server_type, server.name, (count - failed), failed)
-                except Exception as e:
-                    if raise_exception:
-                        raise ValidationError(_("Couldn't get your emails. Check out the error message below for more info:\n%s", e)) from e
+
+            # processing messages in a separate transaction to keep lock on the server
+            server_connection = None
+            message_cr = None
+            try:
+                server_connection = server.connect()
+                message_cr = self.env.registry.cursor()
+                MailThread = server.env['mail.thread'].with_env(self.env(cr=message_cr)).with_context(default_fetchmail_server_id=server.id)
+                thread_process_message = functools.partial(
```

Comment:
Why this ? It makes code harder to read for one usage ?

Reply:
There are multiple reasons.
First, I was trying multiple implementations and the binding was independent of the mail processing loop. Reading `server.attr` in a loop that commits needs to access that record every time, this way, we access it once.